### ISSUE TITLE
Bar Chart not rendering correctly when data is provided to the bar rather than the chart

### ISSUE THREAD

**Issue by FM35** [recharts/recharts#3504]
- [ ] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[https://github.com/recharts/recharts/issues/1699](https://github.com/recharts/recharts/issues/1699)

### Steps to reproduce
https://github.com/recharts/recharts/issues/1699

### What is expected?
https://github.com/recharts/recharts/issues/1699

### What is actually happening?
https://github.com/recharts/recharts/issues/1699

| Environment | Info |
|---|---|
| Recharts | v2.5.0 |
| React | React 18.2.0 |
| System | Ubuntu 21.10 |
| Browser | Mozilla FireFox, 111.0.1 (64-bit) |

**Comment by ckifer** [comment]
Looks like `Bar` doesn't expect the same shape within `data` that `BarChart` does. 

Apparently a lot of the calculations needed for `Bar` are done outside of Bar in `generateCategoricalChart`. This is an architectural problem and probably won't be able to be addressed without some major changes.

**Comment by ckifer** [comment]
Yeah most likely won't be touched for a while. These things are tricky and will require a lot of restructuring which may have unintended side effects.
